### PR TITLE
[Enhancement] Support partition and distribution prune after MV rewrite

### LINKED ISSUE #17931: [Enhancement] Support partition and distribution prune after MV rewrite
## Enhancement

- Support partition and distribution prune after MV rewrite.

### PR DESCRIPTION
**Pull request by LiShuMing**
Signed-off-by: Kevin Li <ming.moriarty@gmail.com>

## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Problem Summary(Required) ：
Fixes #17931
- For each plan after MV rewrite, do some post actions.
```
  /**
     * After plan is rewritten by MV, still do some actions for new MV's plan.
     */
    private List<OptExpression> postRewriteMV(OptimizerContext context, List<OptExpression> candidates) {
        if (candidates == null || candidates.isEmpty()) {
            return Lists.newArrayList();
        }
        List<OptExpression> result = Lists.newArrayList();
        for (OptExpression candidate : candidates) {
            candidate = new MVColumnPruner().pruneColumns(candidate);
            candidate = new MVPartitionPruner().prunePartition(context, candidate);
            result.add(candidate);
        }
        return result;
    }
```
- Keep `mvPartitionPredicate ` which is partition predicates of MV after rewrite, so can be used for `PartitionPrune`.
- To avoid query rewrites multi times, now we restrict one mv is only used once (only query is rewritten we set it used).
- Add `MaterializedViewWithPartitionTest ` to test MV rewrite cases with the partition.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 529 / 589 (89.81%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MVColumnPruner.java|1|2|50.00%|[133]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/OptRemotePartitionPruner.java|123|165|74.55%|[69, 70, 72, 74, 75, 76, 77, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 92, 93, 95, 96, 98, 110, 111, 112, 117, 118, 119, 169, 170, 226, 227, 229, 230, 232, 233, 234, 236, 239, 272]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/OptDistributionPruner.java|18|22|81.82%|[62, 63, 64, 66]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/OptOlapPartitionPruner.java|164|172|95.35%|[253, 278, 279, 281, 292, 303, 304, 306]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java|59|61|96.72%|[587, 606]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|97|100|97.00%|[175, 282, 299]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/RewriteContext.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PartitionPruneRule.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvRewritePreprocessor.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MVPartitionPruner.java|21|21|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateSplit.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MaterializationContext.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/DistributionPruneRule.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RemoteScanPartitionPruneRule.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/rule/BaseMaterializedViewRewriteRule.java|24|24|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MaterializationContext.java [hunk 1]
```diff
@@ -110,4 +112,12 @@ public void setOutputMapping(Map<ColumnRefOperator, ColumnRefOperator> outputMap
     public Set<String> getMvPartitionNamesToRefresh() {
         return mvPartitionNamesToRefresh;
     }
+
+    public boolean isMVUsed() {
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MaterializationContext.java:116]: hasUsed()?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 252 / 259 (97.30%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MVColumnPruner.java|1|2|50.00%|[133]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvRewritePreprocessor.java|26|28|92.86%|[200, 201]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java|59|61|96.72%|[624, 643]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|105|107|98.13%|[180, 300]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/RewriteContext.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MVPartitionPruner.java|31|31|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateSplit.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MaterializationContext.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/rule/BaseMaterializedViewRewriteRule.java|24|24|100.00%|[]|

**Review by Youngwb — DISMISSED** [review]

**Review by Seaven — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java [hunk 2]
```diff
@@ -309,6 +320,23 @@ public static List<ScalarOperator> getAllPredicates(OptExpression root) {
         return predicates;
     }
 
+    public static ScalarOperator getOptExprCompoundPredicate(OptExpression root,
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java:323]:
```suggestion
    public static ScalarOperator rewriteOptExprCompoundPredicate(OptExpression root,
```

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/rule/BaseMaterializedViewRewriteRule.java [hunk 3]
```diff
@@ -61,122 +43,68 @@ protected BaseMaterializedViewRewriteRule(RuleType type, Pattern pattern) {
...
+                mvContext.setMVUsed(true);
+                results.addAll(candidates);
             }
         }
 
         return results;
     }
 
+    /**
+     * After plan is rewritten by MV, still do some actions for new MV's plan.
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/rule/BaseMaterializedViewRewriteRule.java:91]:
```suggestion
     * After plan is rewritten by MV, still do some actions for new MV's plan.
     * 1. column prune
     * 2. partition prune
     * 3. bucket prune
```

File: fe/fe-core/src/test/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MvRewriteOptimizationTest.java [hunk 4]
```diff
@@ -1499,7 +1507,7 @@ public void testUnionRewrite() throws Exception {
 
         String query5 = "select * from multi_mv_1";
         String plan5 = getFragmentPlan(query5);
-        PlanTestBase.assertContains(plan5, "multi_mv_1", "multi_mv_2", "multi_mv_3", "UNION");
+        PlanTestBase.assertContains(plan5, "multi_mv_1", "multi_mv_2", "UNION");
```
Comment (ABingHuang) [inline fe/fe-core/src/test/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MvRewriteOptimizationTest.java:1510]: why the final plan does not contain "multi_mv_3" ?

**Review by ABingHuang — CHANGES_REQUESTED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/rule/BaseMaterializedViewRewriteRule.java [hunk 5]
```diff
@@ -61,122 +43,68 @@ protected BaseMaterializedViewRewriteRule(RuleType type, Pattern pattern) {
...
         }
+        ScalarOperator queryPredicate = MvUtils.getOptExprCompoundPredicate(queryExpression, queryColumnRefRewriter);
         if (!ConstantOperator.TRUE.equals(queryPartitionPredicate)) {
             queryPredicate = MvUtils.canonizePredicate(Utils.compoundAnd(queryPredicate, queryPartitionPredicate));
         }
         final PredicateSplit queryPredicateSplit = PredicateSplit.splitPredicate(queryPredicate);
 
         for (MaterializationContext mvContext : context.getCandidateMvs()) {
+            if (mvContext.hasMVUsed()) {
+                continue;
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/rule/BaseMaterializedViewRewriteRule.java:72]: why add this check?

**Review by ABingHuang — CHANGES_REQUESTED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 242 / 249 (97.19%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MVColumnPruner.java|1|2|50.00%|[133]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvRewritePreprocessor.java|23|25|92.00%|[200, 201]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java|59|61|96.72%|[624, 643]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|105|107|98.13%|[180, 300]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/RewriteContext.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MVPartitionPruner.java|31|31|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateSplit.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/OptOlapPartitionPruner.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/rule/BaseMaterializedViewRewriteRule.java|20|20|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 251 / 256 (98.05%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MVColumnPruner.java|1|2|50.00%|[133]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java|59|61|96.72%|[624, 643]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|105|107|98.13%|[180, 300]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/RewriteContext.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvRewritePreprocessor.java|31|31|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MVPartitionPruner.java|31|31|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateSplit.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/OptOlapPartitionPruner.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/rule/BaseMaterializedViewRewriteRule.java|20|20|100.00%|[]|

File: fe/fe-core/src/test/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MvRewriteOptimizationTest.java (hunk 4 above)
Comment (LiShuMing) [reply fe/fe-core/src/test/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MvRewriteOptimizationTest.java:1510]: `multi_mv_3 ` can be covered by `multi_mv_2`. this is done by creating mv scan operator with partition filters.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MvRewritePreprocessor.java [hunk 6]
```diff
@@ -166,13 +176,46 @@ private LogicalOlapScanOperator createScanMvOperator(MaterializationContext mate
...
+        // - Those predicates are only used for partition/distribution pruning and don't affect the real
+        // query compute.
+        // - TODO: after partition/distribution pruning, those predicates should be removed from mv rewrite result.
+        final OptExpression mvExpression = mvContext.getMvExpression();
+        final List<ScalarOperator> conjuncts = MvUtils.getAllPredicates(mvExpression);
+        final ColumnRefSet mvOutputColumnSet = mvExpression.getOutputColumns();
+        final List<ScalarOperator> mvConjuncts = Lists.newArrayList();
+        // Case1: keeps original predicates which belong to MV table(which are not pruned after mv's partition pruning)
+        for (ScalarOperator conj : conjuncts) {
+            if (conj instanceof BinaryPredicateOperator) {
```
[the remark this answers is not in the document]
Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MvRewritePreprocessor.java]: done

**Review by ABingHuang — APPROVED** [review]
LGTM

**Review by Seaven — APPROVED** [review]

**Comment by LiShuMing** [comment]
run starrocks_admit_test

**Comment by LiShuMing** [comment]
run starrocks_admit_test

**Comment by LiShuMing** [comment]
run starrocks_admit_test

**Comment by LiShuMing** [comment]
run starrocks_admit_test

**Comment by LiShuMing** [comment]
run starrocks_admit_test

**Pull request merged**